Iterative cubic spline

kennethanderson opened the pull request:
### Description

Ruben's idea of fitting a cubic spline to a trajectory is a very good one: #382 .  Forcing the trajectory to a cubic should prevent the controller from overfitting the data, the problem presented here: #160 .  

I've taken the concept a little further by repeatedly fitting the spline and adjusting timestamps until all constraints are satisfied.  This is similar to the way IPTP works, but instead of using local averages, the instantaneous values are used instead.

This technique is presented as another trajectory time parameterization class that can be used instead of IPTP.  It also adds some simple unit tests.  Initial/final velocity and acceleration can be specified.  The constraints imposed are velocity, acceleration, and jerk.  Here is a straight-line trajectory of 11 points:

![spline](https://cloud.githubusercontent.com/assets/13836992/22851156/694ae6e6-efe7-11e6-9a98-c2103960b3b8.png)

Ruben graciously provided some sample trajectories using RRT-Connect with a Kuka IIWA in a fairly cluttered scene. The trajectories have between 25 and 70 waypoints.  These are the resulting spline trajectories (this is the calculated/interpolated spline curves, it is not data from an executing robot or simulation).

![trajectory_1](https://cloud.githubusercontent.com/assets/13836992/22851161/7a9645d0-efe7-11e6-87d4-04fdd8286696.png)

![trajectory_2](https://cloud.githubusercontent.com/assets/13836992/22851162/7a9cb76c-efe7-11e6-817c-f28e53eb4ed7.png)

![trajectory_3](https://cloud.githubusercontent.com/assets/13836992/22851163/7aa4c1aa-efe7-11e6-9036-9fd4365affed.png)

### Checklist
- [x] **Required**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Include a screenshot if changing a GUI
- [ ] Optional: Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [ ] Optional: Decide if this should be cherry-picked to other current ROS branches (Indigo, Jade, Kinetic)

Thank you!

kennethanderson commented:
Ruben, if you could please help with testing, I'd be much obliged.  I haven't worked with robots/ros/moveit in years.

If you plan to compare against 'optimal' methods such as TOPP, you may want to try disabling jerk constraints.  That can be accomplished here:  [https://github.com/kennethanderson/moveit/blob/kinetic-iterative-cubic-spline/moveit_core/trajectory_processing/src/iterative_spline_parameterization.cpp#L72](url)

gavanderhoorn commented:
I've been following this development and this is impressive. The 'staircases' in the jerk plots are very nice to see.

One comment / question: should compile time switches like (en|dis)abling jerk limits (4b692541b5518f1cb4ab6e1f9154f7bd3b2e7f0c) and the default limits (1810abd267c9d906b8d6a2750c97197f13add7e6) all be made configurable / runtime changeable (ultimately through ROS params / dyn reconfig)?

I don't think we can choose a configuration / set of parameters that will suite every usecase or (robotic) setup, so it's probably worth it to make them user configurable.

davetcoleman commented on moveit_ros/planning/planning_request_adapter_plugins/src/add_iterative_spline_parameterization.cpp:
```diff
@@ -0,0 +1,80 @@
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Ken Anderson, based off add_iterative_spline_parameterization.cpp by Ioan Sucan */
```
``add_iterative_spline_parameterization`` should be ``add_time_parameterization``

davetcoleman commented on moveit_ros/planning/planning_request_adapter_plugins/src/add_iterative_spline_parameterization.cpp:
```diff
@@ -0,0 +1,80 @@
+
+namespace default_planner_request_adapters
+{
+class AddSplineParameterization : public planning_request_adapter::PlanningRequestAdapter
```
The name of this class conflicts [with the other version](https://github.com/ros-planning/moveit/pull/382/files#diff-d06ae04150ddd1065b6bae026aa55164) that was recently added by @rbbg to the ``indigo-devel`` branch of moveit. A [PR](https://github.com/ros-planning/moveit/pull/428) exists to cherry-pick it to ``kinetic-devel``, but it has not been merged in.

Since yours is clearly an improvement, I think we should replace his with yours entirely. @rbbg is that ok with you? To make the code history make sense, I think we should still merge https://github.com/ros-planning/moveit/pull/428 then have this PR replace it?

rbbg replied:
That's absolutely fine with me.

davetcoleman replied:
On further thought, merging in #428 and #429 in Jade/Kinetic then syncing with this PR is more trouble than its worth because of slightly different file names. I'd rather just revert @rbbg's PR in Indigo since it was only merged in (prematurely by me) 18 days ago and never announced/documented.

kennethanderson commented:
Yes, you are correct.  I wasn't initially sure if disabling jerk would play nice with the controller problem in #160, but I believe it will.  We just lose initial/final acceleration matching.  

So to that end, I've added arguments to the constructor for jerk enabling and setting the jerk limit.  And I've added a jerk scaling factor similar to the velocity and acceleration scaling factors in computeTimeStamps().  This will allow reducing limit during operation without reconstructing the object.

As for the default limits-- they are really just a fallback if the model doesn't provide them.  This is the same behaviour as IPTP, so I'd prefer to leave that out of this pull request.

kennethanderson commented on moveit_core/trajectory_processing/src/iterative_spline_parameterization.cpp:
```diff
@@ -0,0 +1,510 @@
+};
+
+IterativeSplineParameterization::IterativeSplineParameterization(double max_time_change_per_it, bool add_points)
+  : max_time_change_per_it_(1.0 + max_time_change_per_it), add_points_(add_points)
```
to make it a multiplicative factor.  I changed the member variable name to reflect this.

kennethanderson commented on moveit_core/trajectory_processing/src/iterative_spline_parameterization.cpp:
```diff
@@ -0,0 +1,510 @@
+  double final_acceleration;
+  double max_velocity;
+  double max_acceleration;
+  double max_jerk;
```
Sure.  Adding now.

kennethanderson commented:
@davetcoleman Your requested changes should be complete.  Thanks for reviewing.

rbbg commented:
This is great! I'll definitely try to test the performance as soon as I can. A decent comparison between this and the other available options will be a little more work, but I'd be happy to merge this as soon as it is confirmed that the controller won't do anything weird with this parameterization (which I really doubt).

Seems like `moveit::core::VariableBounds` (and `moveit_msgs/JointLimits`) will also need to be extended in order to fully use and configure jerk limits from the kinematics.yaml :)

rbbg commented on moveit_core/trajectory_processing/src/iterative_spline_parameterization.cpp:
```diff
@@ -0,0 +1,562 @@
+#include <vector>
+
+static const double VLIMIT = 1.0;  // default if not specified in model
+static const double ALIMIT = 3.0;  // default if not specified in model
```
Just wondering: the defaults are currently both set to `1.0`, I don't mind changing that but just wanted to point it out, especially since I've seen quite a number of moveit_configs that don't have the acceleration limit set.

kennethanderson replied:
Good point.  I agree.  Done.

davetcoleman commented:
Waiting on @rbbg's testing

mmoerdijk commented on moveit_ros/planning/planning_request_adapters_plugin_description.xml:
```diff
@@ -30,4 +30,9 @@
     </description>
   </class>
 
+  <class name="default_planner_request_adapters/AddIterativeSplineParameterization" type="default_planner_request_adapters::AddIterativeSplineParameterization" base_class_type="planning_request_adapter::PlanningRequestAdapter">
```
I think this should be:
```
<class name="default_planner_request_adapters/AddSplineParameterization" type="default_planner_request_adapters::AddSplineParameterization" base_class_type="planning_request_adapter::PlanningRequestAdapter">
    <description>
    </description>
  </class>
```
To match the names in `add_iterative_spline_parameterization.cpp`

kennethanderson replied:
Good catch.  I've changed the cpp file to match instead to avoid colliding with #382 .

davetcoleman replied:
I'm in favor or reverting #382 (this [PR](https://github.com/ros-planning/moveit/pull/443)) so you shouldn't need to worry about colliding with it. Still, I like your more detailed naming of using "iterative"

davetcoleman commented:
The clang-format test failed, you introduced trailing whitespace in https://github.com/ros-planning/moveit/pull/441/commits/6182605264764e998400d6ef23c5823fb5cc8b5a

rbbg commented:
I finally got round to doing some testing this. For simple stuff, it seems to at least work, but the trajectories are not as nice as the ones I've seen from the earlier plots. Examples:

![osc_tp](https://cloud.githubusercontent.com/assets/8699378/23164502/9d7f131c-f837-11e6-918c-e8eb835a0ebb.png)

![tp_plot2](https://cloud.githubusercontent.com/assets/8699378/23164641/705a00b2-f838-11e6-8454-61412daf874b.png)

These are just linear motions and should be similar to the first image posted on this PR.

(I am generating these plots by reading JointTrajectory messages from a rosbag and have a python script perform the quintic interpolation. dots are the waypoints)

More difficult problems end up in an error, one example:

`Final acceleration 352009700489716634375032208572468727693985499212628104358753506731293535425976118309558011414013130504346821331194218802322194006883946908550860521230245727986868403904459941731406179223154403971598522255578856325231365769210651491341742808270233188576526336.000000 out of bounds`

I believe the accelerations are supposed to be zero at the initial and final states?
Then, (usually) after a few plans, move_group crashes with 

`terminate called after throwing an instance of 'std::runtime_error'  what():  Time is out of dual 32-bit range`

I've inspected the time_from_start stamps in the trajectory and the last stamps are nonsensical (for instance -2^31 for both secs and nsecs). One note; I have been testing this in docker and while that usually doesn't cause such problems, I think it'd be good if someone with a working kinetic could have a go.

kennethanderson commented:
@rbbg Thanks very much for testing!  You may want to try running with jerk _disabled_ for initial tests.  This will cause the initial/final acceleration to be floating (not to go to zero), but the acceleration should still be within bounds.  

I've added a couple fixes.  One fix checks if the initial velocities/accelerations are set.  That one is most likely to help with the crashes.  And I'll test with your data tomorrow.

rbbg commented:
I had a look with jerk disabled (before ccd3b70) and that did help with the crashes. Did about 10 queries and it managed to solve them all (didn't report any errors) and no more crashes. This is a fairly typical trajectory:

![tp1](https://cloud.githubusercontent.com/assets/8699378/23201298/488bde9c-f8d9-11e6-8e0c-deafe557f9af.png)
better but still some oscillations.

Also saw some stuff like this, where the initial conditions in the trajectory weren't 0, so for some reason you get a strange swoop. Not sure where this is coming from, I was planning from a stationary state.
![tp4](https://cloud.githubusercontent.com/assets/8699378/23201323/64cdcef8-f8d9-11e6-8dc5-71b5b46ff0a1.png)

Finally, a fairly complicated plan:
![tp5](https://cloud.githubusercontent.com/assets/8699378/23201377/92e26ab0-f8d9-11e6-9873-933f4e5a9132.png)
Takes over 30s so that's quite long, but that was caused by the planning.

BrettHemes commented on moveit_core/trajectory_processing/src/iterative_spline_parameterization.cpp:
```diff
@@ -0,0 +1,576 @@
+
+  // Initialize times
+  // 0.01 to prevent divide-by-zero
+  std::vector<double> time_diff(trajectory.getWayPointCount() - 1, 0.01);
```
Setting dt to some arbitrarily small value happens twice, here and at line [498](https://github.com/ros-planning/moveit/pull/441/files#diff-7bfd5dc84e8e4fae3ce118b47c59c3ffR498).  Is this necessary?

Also, I would propose considering a smaller value.  Depending on the controller update rate and trajectory density I think 0.01 could be too large for some cases.  What about something like `std::numeric_limits<double>::epsilon()` here instead?

kennethanderson replied:
Sure.  Done.

BrettHemes commented on moveit_core/trajectory_processing/src/iterative_spline_parameterization.cpp:
```diff
@@ -0,0 +1,576 @@
+      time = (dx / max_velocity);
+    else
+      time = (dx / min_velocity);
+    time += 0.001;  // prevent divide-by-zero
```
See related comment at line [323](https://github.com/ros-planning/moveit/pull/441/files#diff-7bfd5dc84e8e4fae3ce118b47c59c3ffR323)

kennethanderson replied:
Done

mmoerdijk commented:
Yesterday I did some testing in docker with kinetic with `b9ba365` where I got similar results to what @rbbg got with his tests. 

I've also been testing `ccd3b70 ` to do this I ported it to jade and I got much better results, did over a 100 planning actions with random valid joint values non of which resulted in a crash and only a few in invalid trajectories due to post processing (with jerk limits enabled). 

To compare with IPTP I generated 30 random points and planned paths to them. I did this with both time parameterisations methods. This is a typical trajectory with the `max_time_change_per_it` parameter set to 0.001 which gave me better results that the default of 0.01:

**IPTP + iterative spline**
![trajectory_from_moveit_22_02_2017_16_07_48time_diff_001 csv](https://cloud.githubusercontent.com/assets/8997401/23219145/93af7c72-f91e-11e6-961b-3a005d38357b.png)

**IPTP** 
![trajectory_from_moveit_22_02_2017_16_00_39time_diff_001 csv](https://cloud.githubusercontent.com/assets/8997401/23219158/9f667818-f91e-11e6-96ea-ac313ab609dc.png)

Although much better, you can still see the same kind of oscillations that @rbbg mentioned.  

Could someone explain that when the velocities or accelerations go trough zero that happens for all the joints at the same time?

kennethanderson commented:
@rbbg Thanks!  There was still a problem which could have caused the initial/final velocities to be uninitialised.  That is where the 'swoop' at the beginning of the trajectory came from.  Should be fixed now.  

@mmoerdijk Thanks for the results!  Sounds like the big ticket item left is dealing with the oscillations.  I don't have an answer yet, but I have some ideas to test.

rbbg commented:
Just a quick thought; this might also be affected by #416. The initial testing trajectories were generated from indigo, and the fact that trajectories from Jade look a little better also points in that direction.

mmoerdijk commented:
@kennethanderson Thanks a lot for implementing this, it looks really promising.

 I did a quick test and your changes improved the situation for the trajectory I posted yesterday.

It now looks like this
![trajectory_from_moveit_23_02_2017_10_46_04time_diff_001 csv](https://cloud.githubusercontent.com/assets/8997401/23253775/04b70c54-f9b6-11e6-819d-1987a3d5e32f.png)

But for other trajectory it made things worse:
old (`ccd3b70`):
![trajectory_from_moveit_22_02_2017_16_08_16time_diff_001 csv](https://cloud.githubusercontent.com/assets/8997401/23253848/36c876f6-f9b6-11e6-9977-0e01fbd548fc.png)
new(`a426ecc`):
![trajectory_from_moveit_23_02_2017_10_46_31time_diff_001 csv](https://cloud.githubusercontent.com/assets/8997401/23253880/474bbd12-f9b6-11e6-8142-7f4e15c5a496.png)

Overall I cannot say if it is an improvement or not. If you like I could log and provide the corresponding input trajectories.

gavanderhoorn commented:
@rbbg wrote:
> Just a quick thought; this might also be affected by #416. The initial testing trajectories were generated from indigo, and the fact that trajectories from Jade look a little better also points in that direction.

I would definitely test this on Indigo. The 'degraded' planning performance of Kinetic OMPL is not solved ([bitbucket/ompl/ompl#309](https://bitbucket.org/ompl/ompl/issues/309/performance-regression-going-from-ompl)), and it would be best not to convolute things by having the time parameterisation being influenced by that.

kennethanderson commented:
@mmoerdijk -- Yes please, I'd like those trajectories.  Send me an email, my profile has it public for the moment.

130s commented:
There's one approval already. Should we wait for another, given many left comments on this PR?

kennethanderson commented:
@130s Probably want to wait.  Doing a lit review now to try and solve to oscillation problem.

davetcoleman commented:
@kennethanderson awesome, I'm very curious your results.

@BrettHemes presented some thoughts about the oscillation at the ROS-I meeting, perhaps he could remind me what they were

BrettHemes commented:
So I played with both methods for some high-density trajectories and the oscillations seem to be present for both methods.  That said, I don't think the oscillations are an issue unique to the newer spline implementation.  In general the underlying root cause is that the waypoints are fit **and stretched** in orders lower than they are eventually executed at.  The spline approach fixes this issue for the most part when it comes to adhering to limits, however, the discrete time stretches are inherently low-order and non-smooth.  This becomes more apparent as the density of the path (wrt waypoints) increases eventually resulting in an oscillation / inflection point at every waypoint.  I think the problem was less obvious with the original IPTP method due to either the ability to violate acceleration bounds (tended to bunch the oscillations at the beginning/end) OR the heuristic time stretch computations (i.e., findT1/findT2).  The spline method doesn't seem to have these computed time deltas and instead uses a fixed dt.

After talking some with Dave I feel like the issue could be resolved by coming up with a clever way of applying some **smooth** time stretching via polynomial, convolution, or other methods such that the trajectory stretches don't cause acceleration explosions.  Admittedly I don't have a clear ideal on how to do this.  In fact, in the limit so-to-speak perfect time stretching will eventually become equivalent to the time-optimal approaches...  and most likely take just as long from a computational time viewpoint.  Sadly, nothing is free.

BrettHemes commented:
I will work on getting my slides up somewhere from the presentation Dave referenced but here are two images showing a comparison on a very dense (~4 pts/mm) trajectory.

IPTP
![image](https://cloud.githubusercontent.com/assets/9865385/25135573/66913fee-2418-11e7-82fc-b4b3beda3a4d.png)

ISTP
![image](https://cloud.githubusercontent.com/assets/9865385/25135603/7f3cc72a-2418-11e7-94bb-68088b77dc4b.png)

v4hn commented:
@mmoerdijk , @rbbg , @kennethanderson , @BrettHemes 

What is your current view on this issue?
Has there been any progress here that is not documented in the issue?

I assume there is no optimal solution to the acceleration spikes available in the near future.
What do you think is the best way to handle this issue for now (also w.r.t. #443)?

rbbg commented:
I haven't looked at this in a while. I'd probably hold off on #443 until there is actually something better available. It's not perfect the way it is, but at least it is an improvement from what was there before.

kennethanderson commented:
Sorry guys, have not been able to resolve to oscillations problem.  Smoothing is the most promising option, but I've run out of free time to continue working on this for now (children).  Someone may have to either continue this approach or try another method.

kennethanderson commented:
Ping.  This code should be ready for merging.

kennethanderson commented:
@v4hn the code is ready, if you'd like to use it.  It should solve the acceleration spikes issue.  However, it looks like this PR has fallen off everyone's radar.

kennethanderson commented:
Ping.

v4hn commented:
@kennethanderson sorry, this is on my list and I'm interested to see this merged soon.
Sadly I didn't manage to merge before ROSCon...

davetcoleman commented:
I would like to merge this PR today. I have just reviewed all the previous comments as well as the previous spline approach graciously contributed by @rbbg that I merged in a month prior to @kennethanderson improved version in this PR. Here's what I think we should do and the rational:

**[Revert](https://github.com/ros-planning/moveit/pull/443) @rbbg's temporary AddSplineParameterization**
- It does the same thing as @kennethanderson's newer *AddIterativeSplineParameterization* except that ["the acceleration limits are still not satisfied completely"](https://github.com/ros-planning/moveit/pull/382)
- Also from the original PR "One problem remains, with the current starting timestamps, the first waypoint often gets dropped by the controller because it occurs in the past, when this happens, the controller will again apply the quintic interpolation and you get the 'wave' as seen in the figure above."
- Having two plugins do the same thing will just confuse people and require more maintenance

**Merge this PR**
- It has unit tests, has been tested with three different example trajectories, and by many users in this thread
- It obeys the acceleration constraints
- It is written by the same author who wrote our original time parameterizer that has been used for 5 years in all of moveit
- It still will not be the default time paramaterization approach in moveit but rather a better alternative that we can upgrade to in ROS-L or ROS-M

Thanks @rbbg for starting this whole spline effort and @kennethanderson for putting in so much testing and research into a far superior approach for moveit!

v4hn commented:
@davetcoleman I won't have the time to look over this today. Looking forward to see this improved!

gavanderhoorn commented:
@davetcoleman wrote (https://github.com/ros-planning/moveit/pull/441#issuecomment-337484593):
> [..]

I think I can agree with your rationale, and I'm pretty sure @rbbg will agree as well.

@kennethanderson's work is indeed very nice. Not perfect (as he wrote himself), but neither is the current IPTP.

hartmanndennis commented:
Can this be merged soon?

v4hn commented:
I am waiting for @davetcoleman to look at this again and merge it.

davetcoleman commented:
hurray!

gavanderhoorn commented:
Should this be considered for backport to Indigo?

The PR doesn't replace anything, so the current behaviour could be maintained.

v4hn commented:
> Should this be considered for backport to Indigo?

Feel free to file a pull-request. Sounds good to me.

RhysMcK commented:
I have recently been having issues commanding a 6-DOF Kuka robot using the paths generated by moveit's motion planning pipeline. However, every now and again i will receive a slight jerk which will exceed the acceleration and torque limits of my physical robot. I was hoping that this altered method might help, however the trajectories outputs i obtained weren't great.

[blue-Planned= motion planning pipeline output , green-Desired= controller output, red-Actual = IGNORE]
* Moveit!'s Default time Parametrization 
![time_param](https://user-images.githubusercontent.com/15118900/39033625-4ea550ea-44b6-11e8-91c5-15f2dd6bbf73.png)

*Spline Parametrization
![spline_param](https://user-images.githubusercontent.com/15118900/39033623-4b37a836-44b6-11e8-9213-d8bbf1eafb1c.png)

Please note that i didn't capture the exact same trajectory for the two cases above... but regardless the oscillation in the latter method is constant across all trajectories i tried.  Any ideas on why this spline-fitting method is produce particular 'non-ideal' trajectories?

hartmanndennis commented:
@RhysMcK can you try https://github.com/ros-planning/moveit/pull/809?

gavanderhoorn commented:
#809 (or: TOTG) is nice, but iirc, totg is also not jerk limited.

skaruparthi commented:
Hi I am trying to use Iterative Spline parameterization for trajectory smoothing with Trajectory Optimization (TrajOpt) algorithm. As moveit do not support trajopt, I originally created a seperate robot model for moveit and used `trajectory_processing::IterativeParabolicTimeParameterization` to compute timestamps. But, now I want to clean my code and bypass moveit. Can someone suggest how I can use TOTP plugin without moveit.
